### PR TITLE
Pass locals to the Template constructor, remove one layer of caching

### PR DESCRIPTION
**Pull request by tenderlove**
My over all goal is to make `ActionView::Template` a read-only data structure, and this PR is one step along the way.  This PR does two things:

1. Pass locals to the template constructor.  This way we don't need to call `Template#locals=`
2. Uses locals as part of the "details key". This way we can eliminate one layer of template caches

There's a couple things about this patch that I'd like some feedback on.  This change:

  https://github.com/rails/rails/compare/pass-locals-to-template?expand=1#diff-40f2a5bcf9edfd65a8efe42a05c53736R144

I had to do a `details[:locals] || locals.sort` because we have tests that directly call `find_all` on the resolver rather than going through the normal template rendering process.  Do we consider the resolver to be public API?  The class isn't really documented, so I would think "no".  If we do consider it public, I think we should make it private.

Second, I changed `Template#initialize` to [require that locals are passed in](https://github.com/rails/rails/blob/43f14f02c04279a44b0de2bb76071e30d435d3bb/actionview/lib/action_view/template.rb#L139).  I found the exception to be very helpful when debugging.  Should we make it required?  Or just default to an empty array?  If we want to make it required, I'll add a deprecation warning, otherwise just default to empty array (I have no opinion about this).

### CODE REVIEW

File: actionview/lib/action_view/template/resolver.rb [hunk 1]
```diff
@@ -143,14 +140,14 @@ def clear_cache
 
     # Normalizes the arguments and passes it on to find_templates.
     def find_all(name, prefix = nil, partial = false, details = {}, key = nil, locals = [])
-      cached(key, [name, prefix, partial], details, locals) do
-        find_templates(name, prefix, partial, details)
+      cached(key, [name, prefix, partial], details) do
+        find_templates(name, prefix, partial, details, details[:locals] || locals.sort)
```
Comment (matthewd) [inline actionview/lib/action_view/template/resolver.rb]: I won't claim to understand all the layers at play here, but passing information that's not in the cache key into a cached value's calculation makes me pretty nervous. As you're deeper in this code, can you point me to why this is safe? (Or, should we instead be shoving `locals` into `details` before the cache block, when such locals are supplied?)

Comment (tenderlove) [reply actionview/lib/action_view/template/resolver.rb]: > passing information that's not in the cache key into a cached value's calculation makes me pretty nervous.

It is part of the cache key, just in a round-about way (unfortunately).  The details hash is [used to calculate the first `key`](https://github.com/rails/rails/blob/7a3afc9a69842ddf3e0edbdf078f8f3bf3c4b8f7/actionview/lib/action_view/lookup_context.rb#L171-L182).  If the locals change, the return value of `DetailsKey.get(user_details)` will change.  In this function `key` is `details_key`.

**Comment by rafaelfranca** [comment]
> Do we consider the resolver to be public API? The class isn't really documented, so I would think "no". If we do consider it public, I think we should make it private.

Yes, the resolver is part of the public API. José's idea was to make possible to people to write custom resolvers to, for example, read a template from the database (he shows an example of that in his book) We at Shopify have a custom resolver to build a list of all view paths that are possible at boot time. Why do you think it should be private?

**Comment by rafaelfranca** [comment]
To be fair, I think we can change this more quickly than other public APIs we have. It is not like every single developer would write a custom resolver, but I think we should keep possible to write one if they like to. If this API changes without deprecation it would not be the end of the world.

**Review by rafaelfranca — APPROVED** [review]

**Comment by tenderlove** [comment]
@rafaelfranca I'll pick up a copy of José's book.  I can use that to get more context, and maybe we can implement the ideas in the book but with a smaller API

**Pull request closed without merging**